File: packages/scelo-py/src/scelo/_alias.py
"""Column inference: the reason ``sc.triangle(df)`` needs no arguments.

Every tools function accepts explicit column names, but when you leave them
out it looks the columns up here: case- and punctuation-insensitive matches
against the alias lists Scelo IDE uses for its own table suggestions, so
``accident_year`` / ``AY`` / ``Origin Year`` all resolve to the origin column.
"""

from __future__ import annotations

import re
from typing import Dict, Iterable, List, Optional, Sequence

import pandas as pd

_NON_ALNUM = re.compile(r"[^a-z0-9]")


def _lc(s: str) -> str:
    return _NON_ALNUM.sub("", str(s).lower())


# Mirrors packages/scelo-core/src/actuarialTables.ts COLUMN_ALIASES, with the
# extra spellings that show up on real extracts. Order matters: the first alias
# that matches wins, so the canonical name leads each list.
COLUMN_ALIASES: Dict[str, List[str]] = {
    "age": [
        "age", "age_at_entry", "ageatentry", "issue_age", "attained_age", "x", "age_x",
        "age_band", "ageband", "age_last", "age_nearest", "entry_age", "age_years",
    ],
    "qx": ["qx", "q_x", "mortality", "mortality_rate", "death_rate", "q", "prob_death", "rate", "qx_ult"],
    "mx": ["mx", "m_x", "central_rate", "hazard", "mu", "mu_x", "force_of_mortality"],
    "lx": ["lx", "l_x", "lives", "survivors", "l"],
    "deaths": ["deaths", "death", "d", "dx", "actual_deaths", "claims_count", "n_deaths", "died", "events", "actual"],
    "exposure": [
        "exposure", "exposures", "exposed", "exposed_to_risk", "etr", "lives_exposed", "person_years",
        "policy_years", "central_exposure", "initial_exposure", "expo", "e", "ex", "time_at_risk",
    ],
    "expected": ["expected", "expected_deaths", "exp_deaths", "e_deaths", "expected_claims"],
    "origin": [
        "origin", "origin_year", "accident_year", "accidentyear", "ay", "uw_year", "underwriting_year",
        "occurrence_year", "loss_year", "year_of_origin", "cohort", "origin_period", "accident_period",
        "acc_year", "accyear", "uwy", "policy_year",
    ],
    "development": [
        "development", "dev", "development_period", "dev_period", "development_year", "dev_year", "lag",
        "delay", "age_months", "development_lag", "dev_lag", "period", "devyear", "development_months",
    ],
    "payment": [
        "payment_year", "calendar_year", "paid_year", "settlement_year", "transaction_year", "report_year",
        "valuation_year", "cal_year", "calendar_period", "payment_period", "cy",
    ],
    "value": [
        "paid", "incurred", "paid_amount", "incurred_amount", "claims", "claim_amount", "amount", "loss",
        "losses", "payments", "value", "cumulative", "reported", "paid_claims", "incurred_claims",
        "claim", "cost", "severity",
    ],
    "premium": [
        "premium_pp", "premium", "annual_premium", "monthly_premium", "prem", "earned_premium",
        "written_premium", "gwp", "gep", "premiums", "ep",
    ],
    "tenor": ["tenor", "maturity", "term", "years", "year", "t", "maturity_years", "tenor_years"],
    "rate": [
        "rate", "zero_rate", "spot", "spot_rate", "yield", "zero", "swap_rate", "par_rate",
        "interest_rate", "zero_coupon", "spot_yield", "r",
    ],
    "sex": ["sex", "gender", "male_female", "m_f"],
    "policy_term": ["policy_term", "policyterm", "term", "term_years", "policy_term_years", "duration_years"],
    "sum_assured": ["sum_assured", "sumassured", "sa", "face_amount", "face", "benefit", "sum_insured", "si", "coverage"],
    "count": ["count", "policy_count", "policycount", "n", "number", "claim_count", "frequency", "num_claims", "nclaims", "claims_count", "policies"],
    "year": ["year", "calendar_year", "cal_year", "period", "yr"],
    "date": ["date", "as_at", "valuation_date", "effective_date", "start_date", "issue_date"],
    "duration": ["duration", "time", "t", "survival_time", "policy_duration", "tenure"],
    "event": ["event", "status", "died", "death", "claimed", "lapsed", "censored"],
    "group": ["group", "segment", "class", "cohort", "risk_class", "region", "band"],
    "actual": ["actual", "observed", "actual_claims", "actual_deaths", "deaths", "claims"],
}


def find_column(columns: Iterable[str], aliases: Sequence[str]) -> Optional[str]:
    """First column whose normalised name matches one of the aliases, else None."""
    cols = list(columns)
    table = {}
    for c in cols:
        table.setdefault(_lc(c), c)
    for a in aliases:
        hit = table.get(_lc(a))
        if hit is not None:
            return hit
    return None


def infer(df: pd.DataFrame, role: str, explicit: Optional[str] = None, *, required: bool = True,
          exclude: Iterable[str] = ()) -> Optional[str]:
    """Resolve a column for ``role`` ("age", "origin", …).

    ``explicit`` wins when given (and is validated). Otherwise the alias list
    for the role is searched, skipping anything in ``exclude`` (so the same
    column is not picked for two roles). Raises a ``KeyError`` naming the
    role, the aliases tried and the columns available when ``required``.
    """
    if explicit is not None:
        if explicit not in df.columns:
            raise KeyError(f'column "{explicit}" is not in the data (have: {", ".join(map(str, df.columns))})')
        return explicit
    aliases = COLUMN_ALIASES.get(role)
    if aliases is None:
        raise KeyError(f"unknown column role {role!r}")
    ex = set(exclude)
    hit = find_column([c for c in df.columns if c not in ex], aliases)
    if hit is None and required:
        raise KeyError(
            f"could not infer the {role} column: pass {role}=<name>. "
            f"Tried {', '.join(aliases[:6])}…; columns: {', '.join(map(str, df.columns))}"
        )
    return hit


def numeric_columns(df: pd.DataFrame) -> List[str]:
    """Names of the numeric (non-boolean) columns."""
    return [c for c in df.columns if pd.api.types.is_numeric_dtype(df[c]) and not pd.api.types.is_bool_dtype(df[c])]

File: packages/scelo-r/R/alias.R
# Column inference: the reason sc_triangle(df) needs no arguments.
#
# Every tools function accepts explicit column names, but when you leave
# them out it looks the columns up here: case- and punctuation-insensitive
# matches against the alias lists Scelo IDE uses for its own table
# suggestions, so accident_year / AY / Origin Year all resolve to origin.

.sc_lc <- function(s) gsub("[^a-z0-9]", "", tolower(as.character(s)))

#' Column aliases Scelo recognises
#'
#' The alias lists behind column inference, one entry per role ("age",
#' "origin", "development", "value", "premium", ...), canonical name first.
#' @format A named list of character vectors.
#' @export
SC_COLUMN_ALIASES <- list(
  age = c("age", "age_at_entry", "ageatentry", "issue_age", "attained_age", "x", "age_x", "age_band", "ageband", "age_last", "age_nearest", "entry_age", "age_years"),
  qx = c("qx", "q_x", "mortality", "mortality_rate", "death_rate", "q", "prob_death", "rate", "qx_ult"),
  mx = c("mx", "m_x", "central_rate", "hazard", "mu", "mu_x", "force_of_mortality"),
  lx = c("lx", "l_x", "lives", "survivors", "l"),
  deaths = c("deaths", "death", "d", "dx", "actual_deaths", "claims_count", "n_deaths", "died", "events", "actual"),
  exposure = c("exposure", "exposures", "exposed", "exposed_to_risk", "etr", "lives_exposed", "person_years", "policy_years", "central_exposure", "initial_exposure", "expo", "e", "ex", "time_at_risk"),
  expected = c("expected", "expected_deaths", "exp_deaths", "e_deaths", "expected_claims"),
  origin = c("origin", "origin_year", "accident_year", "accidentyear", "ay", "uw_year", "underwriting_year", "occurrence_year", "loss_year", "year_of_origin", "cohort", "origin_period", "accident_period", "acc_year", "accyear", "uwy", "policy_year"),
  development = c("development", "dev", "development_period", "dev_period", "development_year", "dev_year", "lag", "delay", "age_months", "development_lag", "dev_lag", "period", "devyear", "development_months"),
  payment = c("payment_year", "calendar_year", "paid_year", "settlement_year", "transaction_year", "report_year", "valuation_year", "cal_year", "calendar_period", "payment_period", "cy"),
  value = c("paid", "incurred", "paid_amount", "incurred_amount", "claims", "claim_amount", "amount", "loss", "losses", "payments", "value", "cumulative", "reported", "paid_claims", "incurred_claims", "claim", "cost", "severity"),
  premium = c("premium_pp", "premium", "annual_premium", "monthly_premium", "prem", "earned_premium", "written_premium", "gwp", "gep", "premiums", "ep"),
  tenor = c("tenor", "maturity", "term", "years", "year", "t", "maturity_years", "tenor_years"),
  rate = c("rate", "zero_rate", "spot", "spot_rate", "yield", "zero", "swap_rate", "par_rate", "interest_rate", "zero_coupon", "spot_yield", "r"),
  sex = c("sex", "gender", "male_female", "m_f"),
  policy_term = c("policy_term", "policyterm", "term", "term_years", "policy_term_years", "duration_years"),
  sum_assured = c("sum_assured", "sumassured", "sa", "face_amount", "face", "benefit", "sum_insured", "si", "coverage"),
  count = c("count", "policy_count", "policycount", "n", "number", "claim_count", "frequency", "num_claims", "nclaims", "claims_count", "policies"),
  year = c("year", "calendar_year", "cal_year", "period", "yr"),
  date = c("date", "as_at", "valuation_date", "effective_date", "start_date", "issue_date"),
  duration = c("duration", "time", "t", "survival_time", "policy_duration", "tenure"),
  event = c("event", "status", "died", "death", "claimed", "lapsed", "censored"),
  group = c("group", "segment", "class", "cohort", "risk_class", "region", "band"),
  actual = c("actual", "observed", "actual_claims", "actual_deaths", "deaths", "claims")
)

#' Find a column by alias
#'
#' @param columns Column names.
#' @param aliases Candidate spellings (first match wins), compared after
#'   lower-casing and stripping punctuation.
#' @return The matching column name, or `NULL`.
#' @examples
#' sc_find_column(c("Accident Year", "Paid"), SC_COLUMN_ALIASES$origin)
#' @export
sc_find_column <- function(columns, aliases) {
  keys <- .sc_lc(columns)
  for (a in aliases) {
    hit <- match(.sc_lc(a), keys)
    if (!is.na(hit)) return(columns[hit])
  }
  NULL
}

#' Resolve a column for a role
#'
#' `sc_infer(df, "origin")` returns the origin column of `df` by alias;
#' `explicit` wins when given and is validated.
#' @param df A data frame.
#' @param role A name in [SC_COLUMN_ALIASES].
#' @param explicit A column name, or `NULL` to infer.
#' @param required Error (`TRUE`) or return `NULL` when nothing matches.
#' @param exclude Columns not to pick (already used for another role).
#' @return A column name (or `NULL`).
#' @export
sc_infer <- function(df, role, explicit = NULL, required = TRUE, exclude = character()) {
  if (!is.null(explicit)) {
    if (!explicit %in% names(df)) stop(sprintf('column "%s" is not in the data (have: %s)', explicit, paste(names(df), collapse = ", ")), call. = FALSE)
    return(explicit)
  }
  aliases <- SC_COLUMN_ALIASES[[role]]
  if (is.null(aliases)) stop(sprintf("unknown column role '%s'", role), call. = FALSE)
  hit <- sc_find_column(setdiff(names(df), exclude), aliases)
  if (is.null(hit) && required) {
    stop(sprintf("could not infer the %s column: pass %s = <name>. Tried %s...; columns: %s", role, role,
                 paste(utils::head(aliases, 6), collapse = ", "), paste(names(df), collapse = ", ")), call. = FALSE)
  }
  hit
}

.sc_numeric_columns <- function(df) names(df)[vapply(df, function(c) is.numeric(c) && !is.logical(c), logical(1))]
